### DISCUSSION TITLE
Re-useable modules hardfork or add a framework dependency?

### DISCUSSION THREAD

**Discussion by Arkensor** [EveronLife/EveronLife#144] [Polls]
Hey folks, 

I want to bring up the topic of dependencies for the framework. I have mentioned that the next AR update (hopefully this week) includes some powerful script API changes to make our persistence system become very user-friendly by auto-detecting a lot of information out of the box. I think it is going to work great, I am very pleased with the current API draft (not yet fully committed) and I think the system could be re-useable for all kinds of game modes. 
Because BI has helped us by doing 3 major API changes to enable all this, I want not only our project but the AR community overall to benefit from it. Hence I consider releasing the system as a standalone mod. I will need to do docs etc for "our" use case anyway, so it's like one afternoon of effort to make a hard copy and rename the prefixes. Now the question is: Do we want to add this as a dependency to our framework or do you think we should keep it as a hard fork/copy so the core framework is self-contained and does not depend on anything. This question might arise for other systems we design that could benefit others. The code is under MIT so everybody is able to take it if they wanted it for their projects anyway.

**Pro:** 
- Adding dependencies would make it easier for people like me to maintain improvements, so I do not need to port changes I do from EL to standalone or vice versa.
- Reputation (framework and its contributors) for making useful modules useable easier, so they do not need to "rip" out something they want to use.

**Contra:**
- I think I would be annoyed if I have to add like 10 dependencies just to use "vanilla" everon life.
- If the dependency updates the mod might break. (I am not sure if we can do something like with npm where we say, we use mod@version and freeze that)

Let me know your thoughts on this. Do you have any other systems in mind that you think could benefit other game modes?

**Comment by Jaccodouma** [comment]
I'm absolutely all for releasing reusable bits of the framework for others to use. AR should auto download stuff for whatever server you join anyways right so it's not that big an inconvenience for players?

**Comment by thebonbon** [comment]
If we start doing standalone mods, everyone will make their features standalone and we will have not only 10 but 100s of dependencies.
Then you can just piece together your server with every feature you want and not just deactivate it in a config.
-> Not really a problem with autodownload but I think this would rip the framework team apart

I think the Framework should be a single mod that includes everything you need for a basic playable server. Download the mod and done.
Then you can add other mods on top.

**Reply by Arkensor** [reply]
That is a valid concern yeah, I would sort of dislike the nature of everyone baking their own cake and we need to somehow stitch it together

**Reply by Jaccodouma** [reply]
Maybe I've been looking at this project differently. My expectation has been that this will become a framework for communities to implement, meaning everyone **will** (and kinda has to/is expected to) bake their own cake. Then if a community dev makes meaningful changes to the framework they will hopefully choose to contribute that here.

**Comment by 12gabriel3** [comment]
I think it is a good idea. I disagree with thebonbon about it ripping the team apart. If anything it would:

* Attract more contributors. Someone developing another game mode could benefit by contributing to one of our dependencies.
* Make more people use our content.

Restricting the content we make to Life game mode might even encourage contributors to part ways and release their own workshop item apart from Everon Life.

That said, I think we should be careful with what we set as a dependency, core or addon. We need some criteria

**Comment by thebonbon** [comment]
> Do you have any other systems in mind that you think could benefit other game modes?

Most of our system could be used in other gamemodes:

Like logging, farming and the whole food, drink system can be used in any other survival gamemode. (Even the hud)
Money, vehicle shops and like you said persistence can also be used in any other gamemode with currency and shops.
